src/flow_graph.py

```diff
@@ -1096,15 +1274,104 @@ def reset_block_info(self) -> None:
             node.block.block_info = None
 
 
+def locs_clobbered_until_dominator(node: Node) -> Set[Location]:
+    assert node.immediate_dominator is not None
+
+    seen = {node.immediate_dominator}
+    stack = node.parents[:]
+    clobbered = set()
+    while stack:
+        n = stack.pop()
+        if n in seen:
+            continue
+        seen.add(n)
+        for instr in n.block.instructions:
+            clobbered.update(instr.outputs)
+            clobbered.update(instr.clobbers)
+        stack.extend(n.parents)
+    return clobbered
+
+
+def nodes_to_flowgraph(
+    nodes: List[Node], function: Function, arch: ArchFlowGraph
+) -> FlowGraph:
+    flow_graph = FlowGraph(nodes)
+
+    def process_node(node: Node, loc_srcs: LocationRefSetDict) -> None:
+        # Calculate register usage for each instruction in this node
+        for ref in node.block.instruction_refs:
+            inputs = LocationRefSetDict()
+            flow_graph.instr_inputs[ref] = inputs
+            flow_graph.instr_uses[ref] = LocationRefSetDict()
+
+            # Calculate the source of each location
+            for inp in ref.instruction.inputs:
+                sources = RefSet()
+                for loc, srcs in loc_srcs.items():
+                    if locations_alias(loc, inp):
+                        sources.update(srcs)
+                inputs[inp] = sources
+
+            # Remove any clobbered locations
+            for clob in ref.instruction.clobbers + ref.instruction.outputs:
+                for loc in loc_srcs:
+                    if locations_alias(loc, clob) or (
+                        isinstance(loc, StackLocation)
+                        and clob == arch.stack_pointer_reg
+                    ):
+                        loc_srcs.remove(loc)
+                        break
+
+            # Mark outputs as coming from this instruction
+            for out in ref.instruction.outputs:
+                loc_srcs[out] = RefSet([ref])
+
+        # Process nodes dominated by this node, now that we know our own
+        # final Location sources. This will eventually reach every node.
+        for child in node.immediately_dominates:
+            # Create phi refsets for the loc_srcs map for this child
+            child_loc_srcs = loc_srcs.copy()
+            for loc in locs_clobbered_until_dominator(child):
+                child_loc_srcs[loc] = RefSet.phi(loc, child.block)
+
+            process_node(child, child_loc_srcs)
+
+    # Set all registers to come from an unspecified source at the start of the entry node
+    entry_node = flow_graph.entry_node()
+    entry_reg_srcs = LocationRefSetDict()
+    for r in arch.all_regs:
+        entry_reg_srcs.refs[r] = RefSet.phi(r, entry_node.block)
```

Comment:
I mentioned this on Discord, but I was wondering if there was a better term for `PhiRef` that made it clear it also covered inputs to functions? 

It'd actually be OK to use a unique singleton here -- but since I'm using `PhiRef` to indicate the "the initial value of `loc` at the start of `node`" it doesn't seem too wild to also use it here?